src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -0,0 +1,732 @@
+// Return Value:
+// - <none>
+void WindowEmperor::_destroyNotificationIcon()
+{
+    _notificationIcon->RemoveIconFromNotificationArea();
+    _notificationIcon = nullptr;
+}
+
+void WindowEmperor::_checkWindowsForNotificationIcon()
+{
+    // We need to check some conditions to show the notification icon.
+    //
+    // * If there's a Quake window somewhere, we'll want to keep the
+    //   notification icon.
+    // * There's two settings - MinimizeToNotificationArea and
+    //   AlwaysShowNotificationIcon. If either one of them are true, we want to
+    //   make sure there's a notification icon.
+    //
+    // If both are false, we want to remove our icon from the notification area.
+    // When we remove our icon from the notification area, we'll also want to
+    // re-summon any hidden windows, but right now we're not keeping track of
+    // who's hidden, so just summon them all. Tracking the work to do a "summon
+    // all minimized" in GH#10448
+    //
+    // To avoid races between us thinking the settings updated, and the windows
+    // themselves getting the new settings, only ask the app logic for the
+    // RequestsTrayIcon setting value, and combine that with the result of each
+    // window (which won't change during a settings reload).
+    bool needsIcon = _app.Logic().RequestsTrayIcon();
+    {
+        auto windows{ _windows.lock_shared() };
+        for (const auto& _windowThread : *windows)
+        {
+            needsIcon |= _windowThread->Logic().IsQuakeWindow();
+        }
+    }
+
+    if (needsIcon)
+    {
+        _showNotificationIconRequested();
+    }
+    else
+    {
+        _hideNotificationIconRequested();
+    }
+}
+
+void WindowEmperor::_showNotificationIconRequested()
+{
+    if (!_notificationIcon)
+    {
+        _createNotificationIcon();
+    }
+}
+
+void WindowEmperor::_hideNotificationIconRequested()
+{
+    // Destroy it only if our settings allow it
+    if (_notificationIcon)
+    {
+        // If we no longer want the tray icon, but we did have one, then quick
+        // re-summon all our windows, so they don't get lost when the icon
+        // disappears forever.
+        _manager.SummonAllWindows();
+
+        _destroyNotificationIcon();
+    }
+}
+#pragma endregion
+
+// A callback to the window's logic to let us know when the window's
+// quake mode state changes. We'll use this to check if we need to add
+// or remove the notification icon.
+winrt::fire_and_forget WindowEmperor::_windowIsQuakeWindowChanged(winrt::Windows::Foundation::IInspectable sender,
+                                                                  winrt::Windows::Foundation::IInspectable args)
+{
+    co_await wil::resume_foreground(this->_dispatcher);
+    _checkWindowsForNotificationIcon();
+}
+winrt::fire_and_forget WindowEmperor::_windowRequestUpdateSettings()
+{
+    // We MUST be on the main thread to update the settings. We will crash when trying to enumerate fragment extensions otherwise.

```

Comment:
HA! The funny thing is that fragment extension enumeration has to go to great lengths to _not_ crash on the main thread (!) because of awaiting an async. Round and round it goes.